apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/SessionPool.TargetedSessionPool.cs

```diff
@@ -0,0 +1,612 @@
+            private TaskCompletionSource<int> _shutdownTask = null;
+            private bool Shutdown => Interlocked.CompareExchange(ref _shutdownTask, null, null) != null;
+
+            /// <summary>
+            /// The difference between "acquire" and "release" calls, including pending "acquire" calls. Not guarded by the lock,
+            /// but should be modified using Interlocked and accessed via the property.
+            /// </summary>
+            private int _activeSessionCount;
+
+            /// <summary>
+            /// The number of sessions we're currently trying to create. Not guarded by the lock, but should be accessed using Interlocked.
+            /// </summary>
+            private int _inFlightSessionCreationCount;
+
+            /// <summary>
+            /// Thread-safe read-only access to active session count.
+            /// </summary>
+            internal int ActiveSessionCount => Interlocked.CompareExchange(ref _activeSessionCount, 0, 0);
+
+            /// <summary>
+            /// Thread-safe read-only access to in-flight session creation count.
+            /// </summary>
+            internal int InFlightSessionCreationCount => Interlocked.CompareExchange(ref _inFlightSessionCreationCount, 0, 0);
+
+            // Statistics maintained purely for diagnostic purposes. This lets us evaluate
+            // how effective transaction pre-warming is.
+            private long _rwTransactionRequests;
+            private long _rwTransactionRequestsPrewarmed;
+
+            internal TargetedSessionPool(SessionPool parent, DatabaseName databaseName, bool acquireSessionsImmediately) : base(parent)
+            {
+                _databaseName = GaxPreconditions.CheckNotNull(databaseName, nameof(databaseName));
+                _createSessionRequest = new CreateSessionRequest { DatabaseAsDatabaseName = databaseName };
+                if (acquireSessionsImmediately)
+                {
+                    StartAcquisitionTasksIfNecessary();
+                }
+            }
+
+            public async Task<PooledSession> AcquireSessionAsync(TransactionOptions transactionOptions, CancellationToken cancellationToken)
+            {
+                // We may need to start more tasks just acquire a session. We need to try this *before* counting the requested
+                // session as being active. (If we have 99 active sessions at the moment, and nothing in the pool, we want to start a task
+                // to create the session we'll end up acquiring. If we wait until we've counted this session as being active, we won't
+                // start the new task as it'll take us over the limit.)
+                // FIXME: We don't cope with the "Minimum pool size of 0, max active = 100, currently active = 99" case.
+                StartAcquisitionTasksIfNecessary();
+
+                bool success = false;
+                try
+                {
+                    Interlocked.Increment(ref _activeSessionCount);
```

Comment:
There might be a problem with incrementing _activeSessionCount here. In the line below there's a good chance we just enqueu the caller, and at that point, while we await the task (a session to be released or to be acquiered from the server) we are counting this caler both as active session and as pending caller. Or, if instead of enqueuing because the pool is not healthy, we go to create a new session, we are counting this caller both as active session and in flight.
I think the _activeSessionCount needs to be incremented after we actually get the session.
If we do that, there's no need to call `StartAcquisitionTaskIfNecessary` at the beginning of this call, and then my other comment about it being a duplicate in `AcquireSessionImplAsync` is not relevant.

Reply:
Fundamentally this is correct in terms of how we've defined "active sessions" (we've had a call for a session, and it hasn't been released) but it's not good in terms of our action.

We may well need two separate counters here, or we (as you suggest here) only increment once the session has been acquired. That has issues as well, I think - if we're nearly at the maximum number of active sessions and we receive several calls at the same time, we don't want to start so many requests that we end up going over the maximum session count. I'll think on this more over the weekend.